orchagent/vrrporch.cpp

```diff
@@ -0,0 +1,244 @@
+            attr.value.s32 = SAI_ROUTER_INTERFACE_TYPE_SUB_PORT;
+            break;			
+        default:
+            SWSS_LOG_ERROR("Unsupported port type: %d", port.m_type);
+            break;
+    }
+    vmac_attrs.push_back(attr);
+    switch(port.m_type)
+    {
+        case Port::PHY:
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+            attr.value.oid = port.m_port_id;
+            break;
+        case Port::LAG:
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+            attr.value.oid = port.m_lag_id;
+            break;
+        case Port::VLAN:
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_VLAN_ID;
+            attr.value.oid = port.m_vlan_info.vlan_oid;
+            break;
+        case Port::SUBPORT:
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_OUTER_VLAN_ID;
+            attr.value.u16 = port.m_vlan_info.vlan_id;
+            vmac_attrs.push_back(attr);
+            attr.id = SAI_ROUTER_INTERFACE_ATTR_PORT_ID;
+            attr.value.oid = port.m_parent_port_id;
+            break;			
+        default:
+            SWSS_LOG_ERROR("Unsupported port type: %d", port.m_type);
+            break;
+    }
+    port_oid = attr.value.oid;
+    vmac_attrs.push_back(attr);
+    SWSS_LOG_NOTICE("vrrp orch add : port %s, ip %s, vmac %s",request.getKeyString(0).c_str(),ip_pfx.to_string().c_str(),mac.to_string().c_str());
+    sai_route_entry_t unicast_route_entry;
+    unicast_route_entry.switch_id = gSwitchId;
+    unicast_route_entry.vr_id = port.m_vr_id;
+    copy(unicast_route_entry.destination, ip_pfx.getIp());
+    subnet(unicast_route_entry.destination, unicast_route_entry.destination);
+    attr.id = SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID;
+    attr.value.oid = gVirtualRouterId;
+    vmac_attrs.push_back(attr);
+    attr.id = SAI_ROUTER_INTERFACE_ATTR_IS_VIRTUAL;
+    attr.value.booldata =  true;
+    vmac_attrs.push_back(attr);
+    //program VRRPMAC.
+    SWSS_LOG_INFO("vrrp orch add, before create_router_interface : port name %s, rif_id %lx, &rif_id %p",request.getKeyString(0).c_str(),vrrp_rif_id,&vrrp_rif_id);
+    sai_status_t vmac_status = sai_router_intfs_api->create_router_interface(&vrrp_rif_id, gSwitchId, (uint32_t)vmac_attrs.size(), vmac_attrs.data());
+    SWSS_LOG_NOTICE("vrrp orch add, after create_router_interface : port name %s, rif_id %lx",request.getKeyString(0).c_str(),vrrp_rif_id);
+    if (vmac_status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to program Vrrp Mac on interface %s, rv:%d",
+                 port.m_alias.c_str(), vmac_status);
+        throw runtime_error("Failed to program Vrrp Mac on interface.");
+    }
+    //program VIP
+    if (!hasSameIpAddr (request.getKeyString(0),request.getKeyIpPrefix(1)))
+    {
+        attr.id = SAI_ROUTE_ENTRY_ATTR_PACKET_ACTION;
+        attr.value.s32 = SAI_PACKET_ACTION_FORWARD;
+        vip_attrs.push_back(attr);
+        attr.id = SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID;
+        attr.value.oid = vrrp_rif_id;
```

Comment:
@vvbrcm did you test this reverse scenario where traffic is coming from the Internet via Router C?

Reply:
Yes, reverse traffic is tested.

For VRRP the VIP route entry will have VRRP router interface as its nexthop. This is needed to punt the dest IP=VIP packets to CPU.

In your case, do you have subnet route entry? This would be needed to punt any traffic dest to host behind the VRRP master.
Once the host traffic is punted to the CPU, ARP resolution should trigger and an arp entry should be added to H/W.